lib/broadway/producer.ex

```diff
@@ -171,12 +189,55 @@ defmodule Broadway.Producer do
   end
 
   @impl true
+  def handle_info({__MODULE__, :cancel_consumers}, %{rate_limiting: %{} = rate_limiting} = state) do
+    rate_limiting =
+      if :queue.is_empty(rate_limiting.message_buffer) do
+        cancel_consumers(state)
+        rate_limiting
+      else
+        %{rate_limiting | draining?: true}
+      end
+
+    {:noreply, [], %{state | rate_limiting: rate_limiting}}
+  end
+
   def handle_info({__MODULE__, :cancel_consumers}, state) do
-    for from <- state.consumers do
-      send(self(), {:"$gen_producer", from, {:cancel, :shutdown}})
+    cancel_consumers(state)
+    {:noreply, [], state}
+  end
+
+  def handle_info({__MODULE__, :handle_next_demand}, state) do
+    case get_and_update_in(state.rate_limiting.demand_buffer, &:queue.out/1) do
+      {{:value, demand}, state} ->
+        case handle_demand(demand, state) do
+          {:noreply, messages, state} ->
+            schedule_next_handle_demand_if_rate_limiting_open(state)
+            {:noreply, messages, state}
+
+          {:noreply, messages, state, :hibernate} ->
+            schedule_next_handle_demand_if_rate_limiting_open(state)
+            {:noreply, messages, state, :hibernate}
+
+          {:stop, reason, state} ->
+            {:stop, reason, state}
+        end
+
+      {:empty, state} ->
+        {:noreply, [], state}
     end
+  end
 
-    {:noreply, [], state}
+  def handle_info({RateLimiter, :reset_rate_limiting}, state) do
+    state = put_in(state.rate_limiting.state, :open)
+
+    {state, messages} = rate_limit_and_buffer_messages(state)
+
+    # If we're not rate limited after emptying the buffer, we'll forward demand upstream.
+    unless state.rate_limiting.draining? do
```

Comment:
@josevalim if we reset the rate limiting and we're draining, we shouldn't forward demand that we buffered, right?

Reply:
Correct.